worker/uniter/relation/statetracker.go

```diff
@@ -481,3 +488,23 @@ func (r *relationStateTracker) Name(id int) (string, error) {
 	}
 	return relationer.ru.Endpoint().Name, nil
 }
+
+// LocalUnitName returns the name for the local unit.
+func (r *relationStateTracker) LocalUnitName() string {
+	return r.unit.Name()
+}
+
+// LocalUnitAndApplicationLife returns the life values for the local unit and
+// application.
+func (r *relationStateTracker) LocalUnitAndApplicationLife() (life.Value, life.Value, error) {
+	if err := r.unit.Refresh(); err != nil {
```

Comment:
It seems odd to me that we need to call Refresh here. Is the unit actually changing since we started the loop? It feels a bit like we're sneaking a peek at the future, when things like the Life of the unit changing should already be part of the Hook Context. (it was already looking for state-based changes on all the objects it is responsible for.)

Reply:
I bumped into this while doing my initial tests and this is my workaround (IIRC I briefly described this in the commit log).

So basically, removing the unit:
- causes the relation to be removed from the remote
- the above code detects the removal but the local unit has not been marked as dying yet even though it is marked so in the db.

Forcing a refresh works around this issue. It can be verified by removing the refresh and running the first QA scenario where it will mistakingly report that the remote unit is going away.

Reply:
The local unit instance eventually syncs up but that happens after this code runs.